File: src/main/java/io/github/hiiragi283/htms/api/extension/TypedResourceLocation.java
package io.github.hiiragi283.htms.api.extension;

import java.util.Optional;

import net.minecraft.util.ResourceLocation;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import com.github.bsideup.jabel.Desugar;

@Desugar
public record TypedResourceLocation<T> (@NotNull ResourceLocation location, @NotNull Class<T> clazz) {

    public TypedResourceLocation(@NotNull String namespace, @NotNull String path, @NotNull Class<T> clazz) {
        this(new ResourceLocation(namespace, path), clazz);
    }

    @NotNull
    public Optional<T> castOptional(@Nullable Object object) {
        return Optional.ofNullable(cast(object));
    }

    @Nullable
    public T cast(@Nullable Object object) {
        return object != null ? clazz.isInstance(object) ? clazz.cast(object) : null : null;
    }
}

File: src/main/kotlin/io/github/hiiragi283/api/extension/TypedResourceLocation.kt
package io.github.hiiragi283.api.extension

import net.minecraft.util.ResourceLocation

data class TypedResourceLocation<T : Any>(val clazz: Class<T>, val id: ResourceLocation) {
    companion object {
        @JvmStatic
        inline fun <reified T : Any> of(id: ResourceLocation): TypedResourceLocation<T> =
            TypedResourceLocation(T::class.java, id)

        @JvmStatic
        inline fun <reified T : Any> of(namespace: String, path: String): TypedResourceLocation<T> =
            TypedResourceLocation(T::class.java, ResourceLocation(namespace, path))
    }

    @Suppress("UNCHECKED_CAST")
    fun cast(obj: Any?): T? = obj as? T
}